`storage`: set `base_offset` for `index_state` during compaction

WillemKauf opened the pull request:
This value would be incorrectly persisted as `0` during and after compaction.

This means batches indexed in the `index_state` during compaction would have an incorrect `relative_offset` added, and furthermore, the value of `base_offset` would forever be `0` on disk.

Fix this by adding `base_offset` as a required parameter in `index_state::make_empty_index()` and setting the `index_base_offset` when constructing the `copy_data_segment_reducer` for use with self compaction and sliding window compaction.

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v25.1.x
- [X] v24.3.x
- [X] v24.2.x
- [X] v24.1.x

## Release Notes

### Bug Fixes

* Fixes a bug that lead to the `base_offset` in `segment_index`es being persisted as 0 during and after compaction.

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/8b042b28-4049-4473-b215-c3f56aaeb291?refs=530b6c5bef98cfd8dc276a7885cbebd2a7394601&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/8b042b28-4049-4473-b215-c3f56aaeb291/license?refs=530b6c5bef98cfd8dc276a7885cbebd2a7394601&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/b2e01ede-d6c3-4245-86a2-a423c5de8ec6?refs=530b6c5bef98cfd8dc276a7885cbebd2a7394601&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/b2e01ede-d6c3-4245-86a2-a423c5de8ec6/license?refs=530b6c5bef98cfd8dc276a7885cbebd2a7394601&source=prChecksComment)

andrwng commented on src/v/storage/segment_deduplication_utils.cc:
```diff
@@ -247,8 +247,9 @@ ss::future<index_state> deduplicate_segment(
     }
 
-    // restore broker timestamp and clean compact timestamp
+    // restore base offset, broker timestamp and clean compact timestamp
```
>but could heavily reduce the efficacy of seeks using the `segment_index`.

Could you elaborate on this a bit, and maybe give an example? Do you mean there is a case where the index will return an entry that is too low? A dedicated unit test would be helpful to illustrate, if it's not too much trouble

WillemKauf replied:
> Could you elaborate on this a bit, and maybe give an example? 

Fair request, I'll have to come back to this. The only inspection I've really done is eye-balling the code to reason that having a `base_offset` of `0` for all segments is probably really bad for [`find_nearest()`](https://github.com/redpanda-data/redpanda/blob/44d7b62a54c1132dd16d135a40d085d5e75f3062/src/v/storage/index_state.cc#L673-L689) performance.

I'm shocked that in general, this base offset being 0 doesn't affect correctness elsewhere in the code.

> A dedicated unit test would be helpful to illustrate, if it's not too much trouble

Also a reasonable request that I will revisit soon.

andrwng replied:
Just making sure I understand: the description in this PR makes it seem like the problem is that we're seeking somewhere we shouldn't be. But ignoring overflow, that doesn't seem to be the case, in which case seek efficacy seems unaffected -- our deltas are offset by the base_offset, but that seems fine because the base offset is also offset by the same amount. Is that your understanding as well?

WillemKauf replied:
You're correct, I hadn't clocked this originally but if the base offset is set to `0` during compaction then future seeks should also be unaffected, since the base offset is consistently wrong.

I will rewrite the PR cover letter, I don't think any commits needs changing though.

WillemKauf replied:
And yes, to your point, the overflow is the only observable issue we saw fall out of the `base_offset` being 0, luckily.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65258-->
<!--807e6736-eb71-4a74-9b5b-27476bf82deb-->
<!--994c07df-0a94-42af-bf9c-b8183d55d05f-->
<!--75c73e56-8ae2-48a9-97db-e76050f587fc-->
<!--title-->
## Retry command for [Build#65258](https://buildkite.com/redpanda/redpanda/builds/65258)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--543c81ca-c0ae-42b8-ab94-4f35f03f3d42-->
<!--TEST RESULTS-->
<!--9cc40a4a-6890-4127-a318-256b8f70fb4f-->
<!--TEST RESULTS-->
<!--afb62a20-8c8b-49a4-a726-47cadd3a84f6-->
<!--TEST RESULTS-->
<!--a6ce5972-24fc-4a6b-b428-6269d5fca96d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65258">build#65258</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65258#0196836d-3a63-421f-92c8-b8ac0bb44fc4 | FLAKY | 6/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/65258#01968361-4318-47a8-ad01-af36d9f99437 | FLAKY | 14/21 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65258#01968361-431b-40c7-8f10-ea1bea8d0e7f | FLAKY | 9/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65258#01968361-431b-40c7-8f10-ea1bea8d0e7f | FLAKY | 15/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65295">build#65295</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65295#019686ee-c114-46a6-9c4e-2f55b7e0d6e2 | FLAKY | 5/21 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65295#019686ee-c11c-46b5-a928-9f467052f2c7 | FLAKY | 7/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65302">build#65302</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65302#01968717-24a2-4767-8a9f-30cada82b163 | FLAKY | 6/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65347">build#65347</a></summary>


| test_id                                                                                                                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65347#0196890c-3dbd-412c-b169-7c7221289cd6 | FLAKY         | 20/21    |
| rptest.tests.maintenance_test.MaintenanceCycleTest.test_leader_distribution.use_rpk=False                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65347#0196890c-3dbe-48b8-b9da-04b0ee6224c4 | FLAKY         | 20/21    |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65347#01968911-800c-4377-8d72-4d2bdf01544d | FLAKY         | 14/21    |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65347#0196890c-3dbd-412c-b169-7c7221289cd6 | FLAKY         | 20/21    |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=None.allow_gaps_cluster_level=True                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65347#01968911-800d-4aca-9838-4285d5668ce4 | FLAKY         | 20/21    |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsets.test_consumer_group_offsets                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65347#0196890c-3dbe-48b8-b9da-04b0ee6224c4 | FLAKY         | 18/21    |
</details>
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65295-->
<!--0071c2d0-e29b-4d86-a325-a1f3c3071b80-->
<!--89d74cee-684c-4b08-92f9-a120ef5867ed-->
<!--title-->
## Retry command for [Build#65295](https://buildkite.com/redpanda/redpanda/builds/65295)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65302-->
<!--7ed49b61-398e-4284-b640-6bcb1371c93f-->
<!--title-->
## Retry command for [Build#65302](https://buildkite.com/redpanda/redpanda/builds/65302)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names

andrwng commented on src/v/storage/segment_deduplication_utils.cc:
```diff
@@ -247,8 +247,9 @@ ss::future<index_state> deduplicate_segment(
-    // restore broker timestamp and clean compact timestamp
+    // restore base offset, broker timestamp and clean compact timestamp
     auto& new_idx = res.new_idx;
+    new_idx.base_offset = seg->index().base_offset();
```
Don't we need to set this before running the reducer? Won't this result in incorrect offset deltas in the index?

WillemKauf replied:
Oops, of course. Fixed.

WillemKauf committed:
`storage`: set `index_state`'s `base_offset` during compaction

This value would be incorrectly persisted as `0` during and after
compaction.

This means batches indexed in the `index_state` during compaction
would have an incorrect `relative_offset` added, and furthermore,
the value of `base_offset` would forever be `0` on disk.

Fix this by adding `base_offset` as a required parameter in
`index_state::make_empty_index()` and setting the `index_base_offset`
when constructing the `copy_data_segment_reducer` for use with self
compaction and sliding window compaction.

WillemKauf committed:
`storage`: remove hardcoded `32_KiB` for `index_state` indexing

We have a `static` variable defined for this purpose.

WillemKauf committed:
`storage`: add `segment_index::get_index_state()`

Will be used in a future commit for a test.

WillemKauf committed:
`storage`: add `TestSegmentIndexReconstructed` compaction test

WillemKauf commented on src/v/storage/tests/compaction_e2e_test.cc:
```diff
@@ -1476,3 +1477,70 @@ INSTANTIATE_TEST_SUITE_P(
   CompactionFixturePlaceHolderBatchTest,
   ::testing::Bool());
+
+TEST_F(CompactionFixtureTest, TestSegmentIndexReconstructed) {
```
This probably isn't a very good test for comparing the `relative_offset_index` vectors due to the step size of `32_KiB`.

We don't really have a good way to override this in tests...

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/856b0611-c3ac-49b5-a134-c29b3bf280df?refs=530b6c5bef98cfd8dc276a7885cbebd2a7394601&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/856b0611-c3ac-49b5-a134-c29b3bf280df/license?refs=530b6c5bef98cfd8dc276a7885cbebd2a7394601&source=prChecksComment)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65347-->
<!--73da4cc6-c509-4ad6-b385-a6faf8591be6-->
<!--d2c498b8-4e79-4868-a49f-164e36b0b7db-->
<!--005779a8-0642-4b55-9142-f70e0a27f572-->
<!--ee004792-1d0a-40f0-9c55-17af24b525f7-->
<!--title-->
## Retry command for [Build#65347](https://buildkite.com/redpanda/redpanda/builds/65347)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/maintenance_test.py::MaintenanceCycleTest.test_leader_distribution@{"use_rpk":false}
tests/rptest/transactions/consumer_offsets_test.py::VerifyConsumerOffsets.test_consumer_group_offsets
tests/rptest/tests/partition_reassignments_test.py::PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/tiered_storage_pause_test.py::TestTieredStoragePause.test_safe_pause_resume@{"allow_gaps_cluster_level":true,"allow_gaps_topic_level":null}
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_unavailable_nodes
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/maintenance_test.py::MaintenanceCycleTest.test_leader_distribution@{"use_rpk":false}
tests/rptest/transactions/consumer_offsets_test.py::VerifyConsumerOffsets.test_consumer_group_offsets
tests/rptest/tests/partition_reassignments_test.py::PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/tiered_storage_pause_test.py::TestTieredStoragePause.test_safe_pause_resume@{"allow_gaps_cluster_level":true,"allow_gaps_topic_level":null}
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_unavailable_nodes

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x

vbotbuildovich commented:
Oops! Something went wrong.

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14782194338).

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25974-v24.2.x-556 remotes/upstream/v24.2.x
git cherry-pick -x dffa97e16c 1a23958cda 2114ad0c2e 530b6c5bef
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14782193945).

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25974-v24.3.x-172 remotes/upstream/v24.3.x
git cherry-pick -x dffa97e16c 1a23958cda 2114ad0c2e 530b6c5bef
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14782193829).
